src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,456 @@
+
+        // Wind amount is limited by the debt ceiling
+        (uint256 Art,,, uint256 line,) = vat.ilks(ilk);
+        uint256 lineWad = line / RAY; // Round down to always be under the actual limit
+        if (_add(Art, amount) > lineWad) {
+            amount = _sub(lineWad, Art);
+        }
+
+        if (amount == 0) {
+            emit Wind(ilk, 0);
+            return;
+        }
+
+        require(int256(amount) >= 0, "DssDirectDepositHub/overflow");
+
+        vat.slip(ilk, address(pool), int256(amount));
+        vat.frob(ilk, address(pool), address(pool), address(this), int256(amount), int256(amount));
+        // normalized debt == erc20 DAI (Vat rate for this ilk fixed to 1 RAY)
+        daiJoin.exit(address(pool), amount);
+        pool.deposit(amount);
+
+        emit Wind(ilk, amount);
+    }
+
+    function _unwind(bytes32 ilk, D3MPoolLike pool, uint256 supplyReduction, uint256 availableLiquidity, Mode mode, uint256 assetBalance) internal {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why it converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        EndLike end_;
+        uint256 daiDebt;
+        if (mode == Mode.NORMAL) {
+            // Normal mode or module just caged (no culled)
+            // debt is obtained from CDP art
+            (,daiDebt) = vat.urns(ilk, address(pool));
+        } else if (mode == Mode.MODULE_CULLED) {
+            // Module shutdown and culled
+            // debt is obtained from free collateral owned by this contract
+            daiDebt = vat.gem(ilk, address(pool));
+        } else {
+            // MCD caged
+            // debt is obtained from free collateral owned by the End module
+            end_ = end;
+            end_.skim(ilk, address(pool));
+            daiDebt = vat.gem(ilk, address(end_));
+        }
+
+        // Unwind amount is limited by how much:
+        // - max reduction desired
+        // - liquidity available
+        // - gem we have to withdraw
+        // - dai debt tracked in vat (CDP or free)
+        uint256 amount = _min(
+                            _min(
+                                _min(
+                                    supplyReduction,
+                                    availableLiquidity
+                                ),
+                                assetBalance
+                            ),
+                            daiDebt
+                        );
+
+        // Determine the amount of fees to bring back
+        uint256 fees = 0;
+        if (assetBalance > daiDebt) {
+            fees = assetBalance - daiDebt;
+
+            if (_add(amount, fees) > availableLiquidity) {
+                // Don't need safe-math because this is constrained above
+                fees = availableLiquidity - amount;
```

Comment:
I originally wrote this, but I'm looking again and it seems like `fees` can overflow if `amount` > `availableLiquidity ` which is not enforced by the if conditional. Something else will fail in this case, but we should be setting `fees` to 0 in this case.